MH-12958 image-convert WOH

wsmirnow opened the pull request:
With the image-convert WOH you will be able to run encoding operations on media package attachments. The best usage example is the image conversation or modification.

Sponsored by SWITCH.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+      </configurations>
+    </operation>
+
+### Sample Encoding Profile
```
For consistency:

```markdown
## Encoding Profile Example
```

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+    <operation
+      id="image-convert"
+      fail-on-error="true"
+      exception-handler-workflow="error"
```
Keep it simple

```diff
-      fail-on-error="true"
￼-      exception-handler-workflow="error"
```

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+
+
+## Parameter Table
+Source tags and flavors can be used in combination.
```
They always can. But what's the effect if you do? This sentence is not helpful at all.

wsmirnow replied:
I will update this part

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+# Image-Convert Workflow Operation
```
Missing entry in the list of workflow operation handlers

wsmirnow replied:
Thank you, I will add an entry to the list.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+# Image-Convert Workflow Operation
+
+## Description
+With the image-convert workflow operation it is possible to run ffmpeg encoding operations on attachments.
```
If that's the purpose, the woh name does not make any sense. It should be something like `compose-attachment` iinstead.

staubesv replied:
I suggest to adapt the documentation as this WOH is supposed to convert images and  cannot handle "attachments" in general.
It should be fine to say something like" The Image-Convert workflow operation allows source images to be converted into target images with different encoding settings".

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+
+## Description
+With the image-convert workflow operation it is possible to run ffmpeg encoding operations on attachments.
+The best given example is to convert images (as these are attachments) from one format to an other.
```
“One example is the conversion of preview images (as these are attachments)”?

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+With the image-convert workflow operation it is possible to run ffmpeg encoding operations on attachments.
+The best given example is to convert images (as these are attachments) from one format to an other.
+
+This operation expects an attachment as input and creates one attachments as output.
```
one attachment<strike>**s**</strike>

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/image-convert-woh.md:
```diff
@@ -0,0 +1,49 @@
+|**target-flavor***|*/image+converted|Apply these flavor to the output attachments.|EMPTY|
+|**encoding-profiles***|jpeg-player,jpeg-search|A comma separated list of encoding profiles to be applied to each input image.|EMPTY|
+
+The options marked with * are requeried.
```
`*` needs to be escaped

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1194,21 +1194,28 @@ private void validateVideoStream(Job job, Track sourceTrack) throws EncoderExcep
+    if (profileIds == null)
+      throw new IllegalArgumentException("At least one encoding profile must be set");
+
     String[] parameters = new String[2];
```
Maybe:
```java
String[] parameters = {StringUtils.join(profileIds, ','),
                       MediaPackageElementParser.getAsXml(image)};
```

lkiesow replied:
Even better since you convert the array to a list later on anyway:

```java
Liat<String> parameters = Arrays.asList(StringUtils.join(profileIds, ','),
                                        MediaPackageElementParser.getAsXml(image));
```

wsmirnow replied:
Yes, there are some more variants to write same think. But from the performance perspective, an fixed size array will win over the dynamic sized list. I prefer not to change this.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1194,21 +1194,28 @@ private void validateVideoStream(Job job, Track sourceTrack) throws EncoderExcep
+      for (String profileId : profileIds) {
+        final EncodingProfile profile = profileScanner.getProfile(profileId);
+        jobLoad = Math.max(jobLoad, profile.getJobLoad());
+      }
```
Since we can use Java 8 features, what about using
```java
float jobLoad = Arrays.stream(profileIds).map(p -> profileScanner.getProfile(p).getJobLoad()).max();
```

gregorydlogan replied:
That seems like a nice utility function which should live somewhere...

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1222,61 +1229,79 @@ public Job convertImage(Attachment image, String profileId) throws EncoderExcept
-    // Create the encoding engine
-    final EncoderEngine encoderEngine = getEncoderEngine();
+        // Create the encoding engine
+        final EncoderEngine encoderEngine = getEncoderEngine();
```
Since you do not process the files in parallel, why create separate encoder engines?

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1321,9 +1346,10 @@ protected String process(Job job) throws ServiceRegistryException {
           serialized = MediaPackageElementParser.getArrayAsXml(resultingElements);
           break;
         case ImageConversion:
+          String[] encodingProfilesArr = StringUtils.split(arguments.get(0), ',');
```
Just to be safe: Can we properly serialize this (e.g. use Gson)? This should usually not break but why knows when someone starts using `,` in an encoding profile.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/endpoint/ComposerRestService.java:
```diff
@@ -633,7 +633,7 @@ public Response convertImage(@FormParam("sourceImage") String sourceImageXml, @F
     try {
       // Asynchronously convert the specified image
-      Job job = composerService.convertImage((Attachment) sourceImage, profileId);
+      Job job = composerService.convertImage((Attachment) sourceImage, StringUtils.split(profileId, ','));
```
Same here: Proper serialization would be great

wsmirnow replied:
This patch extend an always existing functionality. For this reason I would like to keep it backwards compatible. Changing it to an other format (json or xml) may break thinks.

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,275 @@
+  /** The configuration options for this handler */
+  private static final SortedMap<String, String> CONFIG_OPTIONS;
+
+  static {
```
This is unnecessary by now, isn't it?

wsmirnow replied:
What do you mean?

lkiesow replied:
There was a discussion about the `CONFIG_OPTIONS` a while ago. They are basically obsolete and never used anywhere. Half of the workflow operation handlers do not even implement them. Hence. while this is not breaking anything, it's basically just dead code and can be removed unless you are using it for something.

wsmirnow replied:
Ok, I will change this.

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,275 @@
+    Map<Job, Attachment> jobs = new Hashtable<>();
+    try {
+      for (Attachment sourceElement : sourceElements) {
+        Job job = composerService.convertImage(sourceElement, profiles.toArray(new String[profiles.size()]));
```
You are always converting lists to arrays and back. Why not always use a list?

wsmirnow replied:
Because the API must be changed otherwise. I would unlikely do that.

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,275 @@
+      }
+      return createResult(mediaPackage, WorkflowOperationResult.Action.CONTINUE);
+    } catch (WorkflowOperationException ex) {
+      throw ex;
```
Why catch this to simply throw it again?

wsmirnow replied:
Because this method can only throw `WorkflowOperationException`s. At this point it is currently an `WorkflowOperationException` so we do not need to wrap it like the other exceptions. Yes, it is also possible to write
```
catch (Throwable t) {
      if (t instanceof WorkflowOperationException)
        throw (WorkflowOperationException) t;
      else
        throw new WorkflowOperationException(..., t);
}
```

lkiesow replied:
I missed that you are using this as an exception to catching `Throwable`. That makes sense. Though I'm wondering if catching `Throwable` does. Does it make sense to convert system `Error` to `Exception`?

lkiesow commented on modules/composer-workflowoperation/src/test/resources/image_convert_mediapackage.xml:
```diff
@@ -0,0 +1,20 @@
+			<ns2:url>image.png</ns2:url>
+		</ns2:attachment>
+	</ns2:attachments>
+</ns2:mediapackage>
```
tabs and newline at eof

staubesv commented:
Hi @wsmirnow Please note that https://github.com/opencast/opencast/pull/314 has caused conflicts in this PR.

lkiesow commented:
Hi @staubesv, #314 did not touch the `pom.xml`.

staubesv commented:
@wsmirnow & @lkiesow Oh, the conflicts are caused by #299. Anyway, please resolve the merge conflicts.

wsmirnow committed:
MH-12958 image-convert WOH added

The Image-Convert workflow operation allows source images to be converted into target images with different encoding settings.

wsmirnow commented:
Hi @staubesv,
the merge conflict is fixed and the code is rebased. Please continue with the review.

Thanks

gregorydlogan commented:
@wsmirnow This PR does not compile when merged into the latest develop, likely due to changes from #208.  Can you merge develop into this branch please?

wsmirnow committed:
MH-12958 composer service after rebase fixes

wsmirnow committed:
MH-12958 image convert operation should set the mimetype on the output images

wsmirnow commented:
Hi @gregorydlogan,
the PR is rebased on develop. Please continue with the review.

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1235,9 +1242,13 @@ public Attachment convertImageSync(Attachment image, String profileId) throws En
+      List<Attachment> result = convertImage(job, image, profileId);
+      if (result.isEmpty()) {
+        job.setStatus(Job.Status.FAILED);
+        return null;
```
This will cause a NPE at ThumbnailImpl:385 and ComposerRestService:846, although the previous implementation's line 1240 (result.getOrElseNull) would do the same.  Regardless, this should be fixed IMO.

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1252,62 +1263,84 @@ public Attachment convertImageSync(Attachment image, String profileId) throws En
+        } catch (NotFoundException e) {
+          incident().recordFailure(job, WORKSPACE_GET_NOT_FOUND, e,
+                  getWorkspaceMediapackageParams("source image", sourceImage), NO_DETAILS);
+          throw new EncoderException("Requested video track " + sourceImage + " was not found", e);
```
Error description incorrect, should be an image attachment rather than video track I'm assuming?

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1252,62 +1263,84 @@ public Attachment convertImageSync(Attachment image, String profileId) throws En
+        } catch (IOException e) {
+          incident().recordFailure(job, WORKSPACE_GET_IO_EXCEPTION, e,
+                  getWorkspaceMediapackageParams("source image", sourceImage), NO_DETAILS);
+          throw new EncoderException("Error accessing video track " + sourceImage, e);
```
See above.

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1252,62 +1263,84 @@ public Attachment convertImageSync(Attachment image, String profileId) throws En
+        // encoding did not return a file
+        if (!output.exists() || output.length() == 0)
+          throw new EncoderException(format(
+              "Image conversion job %d doesn't created an output file for the source image %s with encoding profile %s",
```
doesn't -> failed to.  Or didn't, either one :)

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,248 @@
+
+  @Override
+  public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobContext context) throws WorkflowOperationException {
+    WorkflowOperationInstance operation = workflowInstance.getCurrentOperation();
```
Generally these config keys are put into constants, and a static is defined containing the CONFIG_OPTIONS TreeMap.  Take a look at CoverImageWorkflowOperationHandler to see what this looks like.

wsmirnow replied:
That is true. We'd collected all the WOH options with their description in a map previously, because the `AbstractWorkflowOperationHandler` had a geta for that map. This was removed by Lars during OC6 development i guess.

I will move the config keys to the static final class members and put a documentation for each one.

lkiesow replied:
@gregorydlogan, the map is not used by anything right now and it's not at all or not fully (options were added but not to the map) implemented by a lot of our workflow operations. Somewhere on my todo list is the item to remove that piece of code entirely since I thought no one cared about that (I brought that to list a while ago). Do you want to keep it for any reason?

lkiesow replied:
Though I agree that using constants instead of inline strings would be nice ;-)

gregorydlogan replied:
I thought it was still in use.  If it's not in use then we can omit it here and should plan to remove it elsewhere :)

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,248 @@
+        jobs.put(job, sourceElement);
+      }
+      if (!waitForStatus(jobs.keySet().toArray(new Job[jobs.size()])).isSuccess()) {
+        throw new WorkflowOperationException("At least one image conversation job does not succeeded.");
```
did not succeed.

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,248 @@
+          List<String> removingTags = new ArrayList<>();
+          for (String targetTag : asList(targetTagsOption)) {
+            if (!StringUtils.startsWithAny(targetTag, "+", "-")) {
+              if (additionalTags.size() > 0 || removingTags.size() > 0) {
```
I would move this logic up into the WOH startup rather than waiting until after the jobs have completed to check this.  The behaviour is undefined at this point - if I were an admin reading this I would be unsure of what's going to happen.  IMO it's better to check this at the beginning, then throw an exception in this case.

wsmirnow replied:
Agree, this logic should be unified on all workflow handlers IMO. But not with this PR.

wsmirnow committed:
MH-12985 pr review comments solved
